From 65999c2360b903f1f0edc4b35954dd6e51bd9975 Mon Sep 17 00:00:00 2001
From: nagachika <nagachika@b2dd03c8-39d4-4d8f-98ff-823fe69b080e>
Date: Sun, 2 Mar 2014 16:30:04 +0000
Subject: [PATCH 4/5] merge revision(s) r45178,r45179,r45180,r45183: [Backport
 #9568]

	eval.c: remove unneeded GC guard

	* eval.c (setup_exception): remove RB_GC_GUARD which is no longer
  needed since r41598.
	* eval.c (setup_exception): preserve errinfo across calling #to_s
	  method on the exception.  [ruby-core:61091] [Bug #9568]


git-svn-id: svn+ssh://ci.ruby-lang.org/ruby/branches/ruby_2_0_0@45251 b2dd03c8-39d4-4d8f-98ff-823fe69b080e
---
 ChangeLog                   |  5 +++++
 eval.c                      |  5 ++++-
 test/ruby/envutil.rb        | 16 ++++++++++++++++
 test/ruby/test_exception.rb | 22 ++++++++++++++++++++++
 test/ruby/test_module.rb    | 10 ++--------
 version.h                   |  2 +-
 6 files changed, 50 insertions(+), 10 deletions(-)

diff --git a/ChangeLog b/ChangeLog
index f78a5f4..8a22461 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+Mon Mar  3 01:25:28 2014  Nobuyoshi Nakada  <nobu@ruby-lang.org>
+
+	* eval.c (setup_exception): preserve errinfo across calling #to_s
+	  method on the exception.  [ruby-core:61091] [Bug #9568]
+
 Mon Mar  3 01:00:00 2014  Nobuyoshi Nakada  <nobu@ruby-lang.org>
 
 	* numeric.c (ruby_num_interval_step_size): check signs and get rid
diff --git a/eval.c b/eval.c
index b5f41e6..e8dc008 100644
--- a/eval.c
+++ b/eval.c
@@ -467,9 +467,12 @@ setup_exception(rb_thread_t *th, int tag, volatile VALUE mesg)
 	!rb_obj_is_kind_of(e, rb_eSystemExit)) {
 	int status;
 
+	mesg = e;
 	PUSH_TAG();
 	if ((status = EXEC_TAG()) == 0) {
-	    RB_GC_GUARD(e) = rb_obj_as_string(e);
+	    th->errinfo = Qnil;
+	    e = rb_obj_as_string(mesg);
+	    th->errinfo = mesg;
 	    if (file && line) {
 		warn_printf("Exception `%s' at %s:%d - %s\n",
 			    rb_obj_classname(th->errinfo),
diff --git a/test/ruby/envutil.rb b/test/ruby/envutil.rb
index faf0dc2..030292f 100644
--- a/test/ruby/envutil.rb
+++ b/test/ruby/envutil.rb
@@ -150,6 +150,22 @@ module EnvUtil
     $VERBOSE = verbose
   end
   module_function :with_default_internal
+
+  def labeled_module(name, &block)
+    Module.new do
+      singleton_class.class_eval {define_method(:to_s) {name}; alias inspect to_s}
+      class_eval(&block) if block
+    end
+  end
+  module_function :labeled_module
+
+  def labeled_class(name, superclass = Object, &block)
+    Class.new(superclass) do
+      singleton_class.class_eval {define_method(:to_s) {name}; alias inspect to_s}
+      class_eval(&block) if block
+    end
+  end
+  module_function :labeled_class
 end
 
 module Test
diff --git a/test/ruby/test_exception.rb b/test/ruby/test_exception.rb
index 107d80b..123eecc 100644
--- a/test/ruby/test_exception.rb
+++ b/test/ruby/test_exception.rb
@@ -80,6 +80,28 @@ class TestException < Test::Unit::TestCase
     assert(!bad)
   end
 
+  def test_errinfo_in_debug
+    bug9568 = EnvUtil.labeled_class("[ruby-core:61091] [Bug #9568]", RuntimeError) do
+      def to_s
+        require '\0'
+      rescue LoadError
+        self.class.to_s
+      end
+    end
+
+    err = EnvUtil.verbose_warning do
+      assert_raise(bug9568) do
+        $DEBUG, debug = true, $DEBUG
+        begin
+          raise bug9568
+        ensure
+          $DEBUG = debug
+        end
+      end
+    end
+    assert_include(err, bug9568.to_s)
+  end
+
   def test_break_ensure
     bad = true
     while true
diff --git a/test/ruby/test_module.rb b/test/ruby/test_module.rb
index 047f7c9..c6979d8 100644
--- a/test/ruby/test_module.rb
+++ b/test/ruby/test_module.rb
@@ -1500,17 +1500,11 @@ class TestModule < Test::Unit::TestCase
   end
 
   def labeled_module(name, &block)
-    Module.new do
-      singleton_class.class_eval {define_method(:to_s) {name}; alias inspect to_s}
-      class_eval(&block) if block
-    end
+    EnvUtil.labeled_module(name, &block)
   end
 
   def labeled_class(name, superclass = Object, &block)
-    Class.new(superclass) do
-      singleton_class.class_eval {define_method(:to_s) {name}; alias inspect to_s}
-      class_eval(&block) if block
-    end
+    EnvUtil.labeled_class(name, superclass, &block)
   end
 
   def test_prepend_instance_methods_false
diff --git a/version.h b/version.h
index dd52021..5bda012 100644
--- a/version.h
+++ b/version.h
@@ -1,6 +1,6 @@
 #define RUBY_VERSION "2.0.0"
 #define RUBY_RELEASE_DATE "2014-03-03"
-#define RUBY_PATCHLEVEL 454
+#define RUBY_PATCHLEVEL 455
 
 #define RUBY_RELEASE_YEAR 2014
 #define RUBY_RELEASE_MONTH 3
-- 
1.9.0

